sw-emulator/lib/cpu/src/cpu.rs

```diff
@@ -408,27 +412,41 @@ impl<TBus: Bus> Cpu<TBus> {
         for action_type in fired_action_types.iter() {
             match action_type {
                 TimerAction::WarmReset => {
+                    self.halted = false;
                     self.reset_pc();
                     break;
                 }
                 TimerAction::UpdateReset => {
+                    self.halted = false;
                     self.reset_pc();
                     break;
                 }
-                TimerAction::Nmi { mcause } => return self.handle_nmi(*mcause, 0),
+                TimerAction::Nmi { mcause } => {
+                    self.halted = false;
+                    return self.handle_nmi(*mcause, 0);
+                }
                 TimerAction::SetNmiVec { addr } => self.nmivec = *addr,
-                TimerAction::ExtInt { irq } => {
-                    if self.global_int_en && self.ext_int_en {
+                TimerAction::ExtInt { irq, can_wake } => {
+                    if self.global_int_en && self.ext_int_en && (!self.halted || *can_wake) {
+                        self.halted = false;
                         return self.handle_external_int(*irq);
                     }
                 }
                 TimerAction::SetExtIntVec { addr } => self.ext_int_vec = *addr,
                 TimerAction::SetGlobalIntEn { en } => self.global_int_en = *en,
                 TimerAction::SetExtIntEn { en } => self.ext_int_en = *en,
+                TimerAction::Halt => self.halted = true,
                 _ => {}
             }
         }
 
+        // We are in a halted state. Don't continue executing but poll the bus for interrupts
+        if self.halted {
+            self.clock.timer().schedule_poll_in(1);
```

Comment:
For efficiency reasons, this isn't how this was intended to work. Peripherals should schedule their own polls for actions that will fire interrupts.

For example, a timer peripheral might schedule a poll when the timer expires, and then when polled, fire the interrupt.

Reply:
I think this might still work if you delete this line.

Reply:
Done. Previous iterrations didn't schedule a poll inside the mailbox, but now that's the case.